lib/assets/data_frame_cell/main.js

```diff
@@ -0,0 +1,571 @@
+          ],
+        },
+        columnTypes: {
+          float: "numeric",
+          integer: "numeric",
+          date: "numeric",
+          string: "categorical",
+          boolean: "boolean",
+        },
+      };
+    },
+
+    computed: {
+      dataFrameInfo() {
+        const dataFrameVariable = this.rootFields.data_frame;
+        const dataFrame = this.dataFrames.find(
+          (df) => df["variable"] === dataFrameVariable
+        );
+        return dataFrame;
+      },
+      dataFrameColumns() {
+        const dataFrame = this.dataFrameInfo;
+        return dataFrame ? Object.keys(dataFrame["columns"]) : [];
+      },
+      noDataFrame() {
+        return !this.rootFields.data_frame;
+      },
+    },
+
+    methods: {
+      handleFieldChange(event) {
+        const field = event.target.name;
+        const idx = event.target.getAttribute("index");
+        const operation = event.target.getAttribute("operation");
+        const value = idx
+          ? this.operations[operation][idx][field]
+          : this.rootFields[field];
+        ctx.pushEvent("update_field", {
+          operation,
+          field,
+          value,
+          idx: idx && parseInt(idx),
+        });
+      },
+      addOperation(operation) {
+        ctx.pushEvent("add_operation", { operation });
+      },
+      removeOperation(operation, idx) {
+        ctx.pushEvent("remove_operation", { operation, idx: parseInt(idx) });
+      },
+      hasOperations(operation) {
+        return this.operations[operation].length > 0;
+      },
+      filterOptionsByType(column) {
+        const columnType = this.dataFrameInfo?.columns[column];
+        return this.filterOptions[this.columnTypes[columnType]];
+      },
+      columnType(column) {
+        return this.dataFrameInfo?.columns[column];
+      },
+    },
+  }).mount(ctx.root);
+
+  ctx.handleEvent("update_root", ({ fields }) => {
+    setRootValues(fields);
+  });
+
+  ctx.handleEvent("update_operation", ({ operation, idx, fields }) => {
+    setOperationValues(operation, idx, fields);
+  });
+
+  ctx.handleEvent("update_data_frame", ({ fields }) => {
+    setOperations(fields.operations);
+    setRootValues(fields.root_fields);
+  });
+
+  ctx.handleEvent("set_available_data", ({ data_options, fields, require }) => {
+    app.dataFrameVariables = data_options.map((df) => df["variable"]);
+    app.dataFrames = data_options;
+    app.missingRequire = require;
+    setOperations(fields.operations);
+    setRootValues(fields.root_fields);
+  });
+
+  ctx.handleEvent("set_operations", (operations) => {
+    setOperations(operations);
+  });
+
+  Vue.watch(
+    () => app.operations.filters,
+    (currentFilters) => {
+      currentFilters.forEach((filter) => {
+        const filterType = app.columnType(filter.column);
+        const numeric = ["integer", "float", "date", "datetime"];
+        if (numeric.includes(filterType)) {
+          filter.value = filter.value?.toString().replace(/[^\d.-]/g, "");
```

Comment:
Why do we need to watch the filters? We know at which events those values change, so we should be able to handle it there, no?

By watching we unnecessarily make passes whenever we get new filters from the server, especially if using with multiple tabs/users.

Reply:
I was doing some real time validations to avoid filter problems, but yes, doesn't worth it. The error messages are enough